File: front/src/services/planner.service.ts
import {BundleSubjectPart} from '../app/shared/model/planner/bundle-subject-part.interface';
import {Injectable} from '@angular/core';
import {Response} from '@angular/http';
import {HttpInterceptorService} from '@covalent/http';
import {Observable} from 'rxjs';
import {environment} from '../environments/environment';
import {Program} from '../app/shared/model/planner/program.interface';
import {Faculty} from '../app/shared/model/planner/faculty.interface';
import {Course} from '../app/shared/model/planner/course.interface';
import {Cohort} from '../app/shared/model/planner/cohort.interface';
import {Subject} from '../app/shared/model/planner/subject.interface';
import {SingleSubject} from '../app/shared/model/planner/single-subject.interface';
import {BundleSubject} from '../app/shared/model/planner/bundle-subject.interface';
import {AcademicSession} from '../app/shared/model/planner/academic-session.interface';
import {AcademicYear} from '../app/shared/model/planner/academic-year.interface';
import {ProgramLevel} from '../app/shared/model/planner/program-level.interface';
import {Curriculum} from '../app/shared/model/planner/curriculum.interface';

@Injectable()
export class PlannerService {

  private PLANNER_API: string = environment.endpoint + '/api/planner';

  constructor(private _http: HttpInterceptorService) {
  }

  // ====================================================================================================
  // ACADEMIC YEAR
  // ====================================================================================================

  findAcademicYears(): Observable<AcademicYear[]> {
    return this._http.get(this.PLANNER_API + '/academicYears')
      .map((res: Response) => <AcademicYear[]>res.json());
  }

  // ====================================================================================================
  // ACADEMIC SESSION
  // ====================================================================================================

  findAcademicSessions(): Observable<AcademicSession[]> {
    return this._http.get(this.PLANNER_API + '/academicSessions')
      .map((res: Response) => <AcademicSession[]>res.json());
  }

  findAcademicSessionByCode(code: string): Observable<AcademicSession> {
    return this._http.get(this.PLANNER_API + '/academicSessions/' + code)
      .map((res: Response) => <AcademicSession>res.json());
  }

  activateAcademicSession(academicSession: AcademicSession): Observable<String> {
    console.log('activate academic session');
    return this._http.get(this.PLANNER_API + '/academicSessions/' + academicSession.code + '/activate')
      .map((res: Response) => <String>res.json());
  }

  deactivateAcademicSession(academicSession: AcademicSession): Observable<String> {
    console.log('deactivate academic session');
    return this._http.get(this.PLANNER_API + '/academicSessions/' + academicSession.code + '/deactivate')
      .map((res: Response) => <String>res.json());
  }

  saveAcademicSession(academicSession: AcademicSession): Observable<String> {
    return this._http.post(this.PLANNER_API + '/academicSessions/' + academicSession.code + '/save', JSON.stringify(academicSession))
      .flatMap((res: Response) => Observable.of(res.text()))
      .catch((error) => this.handleError(error));
  }

  updateAcademicSession(academicSession: AcademicSession): Observable<String> {
    return this._http.put(this.PLANNER_API + '/academicSessions/' + academicSession.code + '/update', JSON.stringify(academicSession))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // FACULTY
  // ====================================================================================================

  findFaculties(): Observable<Faculty[]> {
    console.log('findFaculties');
    return this._http.get(this.PLANNER_API + '/faculties')
      .map((res) => {
        return <Faculty[]>res.json();
      });
  }

  findFacultyByCode(code: string): Observable<Faculty> {
    console.log('findFacultyByCode');
    return this._http.get(this.PLANNER_API + '/faculties/' + code)
      .map((res: Response) => <Faculty>res.json());
  }

  activateFaculty(faculty: Faculty): Observable<String> {
    console.log('activate Faculty');
    return this._http.get(this.PLANNER_API + '/faculties/' + faculty.code + '/activate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  deactivateFaculty(faculty: Faculty): Observable<String> {
    console.log('deactivate Faculty');
    return this._http.get(this.PLANNER_API + '/faculties/' + faculty.code + '/deactivate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  saveFaculty(faculty: Faculty): Observable<String> {
    return this._http.post(this.PLANNER_API + '/faculties/' + faculty.code + '/save', JSON.stringify(faculty))
      .flatMap((res: Response) => Observable.of(res.text()))
      .catch((error) => this.handleError(error));
  }

  updateFaculty(faculty: Faculty): Observable<String> {
    return this._http.put(this.PLANNER_API + '/faculties/' + faculty.code, JSON.stringify(faculty))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // PROGRAM
  // ====================================================================================================

  findPrograms(): Observable<Program[]> {
    console.log('findPrograms');
    return this._http.get(this.PLANNER_API + '/programs')
      .map((res: Response) => <Program[]>res.json());
  }

  findProgramByCode(code: string): Observable<Program> {
    return this._http.get(this.PLANNER_API + '/programs/' + code)
      .map((res: Response) => <Program>res.json());
  }

  saveProgram(program: Program): Observable<String> {
    return this._http.post(this.PLANNER_API + '/programs/' + program.code + '/save', JSON.stringify(program))
      .flatMap((res: Response) => Observable.of(res.text()));
      //.catch((error) => this.handleError(error));
  }

  updateProgram(program: Program): Observable<String> {
    return this._http.put(this.PLANNER_API + '/programs/' + program.code, JSON.stringify(program))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeProgram(program: Program): Observable<Program> {
    return this._http.delete(this.PLANNER_API + '/programs/' + program.code)
      .map((res: Response) => <Program>res.json());
  }

  activateProgram(program: Program): Observable<String> {
    console.log('activate program');
    return this._http.get(this.PLANNER_API + '/programs/' + program.code + '/activate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  deactivateProgram(program: Program): Observable<String> {
    console.log('deactivate program');
    return this._http.get(this.PLANNER_API + '/programs/' + program.code + '/deactivate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // COURSE
  // ====================================================================================================

  findCourses(): Observable<Course[]> {
    console.log('findCourses');
    return this._http.get(this.PLANNER_API + '/courses')
      .map((res: Response) => <Course[]>res.json());
  }

  findCourseByCode(code: string): Observable<Course> {
    return this._http.get(this.PLANNER_API + '/courses/' + code)
      .map((res: Response) => <Course>res.json());
  }

  saveCourse(course: Course): Observable<String> {
    console.log(course.code);
    return this._http.post(this.PLANNER_API + '/courses/' + course.code + '/save', JSON.stringify(course))
      .flatMap((res: Response) => Observable.of(res.text()))
      .catch((error) => this.handleError(error));
  }

  addCourse(course: Course): Observable<String> {
    return this._http.post(this.PLANNER_API + '/courses/' + course.code + '/add', JSON.stringify(course))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateCourse(course: Course): Observable<String> {
    return this._http.put(this.PLANNER_API + '/courses/' + course.code, JSON.stringify(course))
      .flatMap((res: Response) => Observable.of(res.text()));

  }

  removeCourse(course: Course): Observable<Course> {
    return this._http.delete(this.PLANNER_API + '/course/' + course.code)
      .map((res: Response) => <Course>res.json());
  }

  activateCourse(course: Course): Observable<String> {
    console.log('activate course');
    return this._http.get(this.PLANNER_API + '/courses/' + course.code + '/activate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  deactivateCourse(course: Course): Observable<String> {
    console.log('deactivate course');
    return this._http.get(this.PLANNER_API + '/courses/' + course.code + '/deactivate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // COHORT
  // ====================================================================================================

  findCohorts(): Observable<Cohort[]> {
    console.log('findCohorts');
    return this._http.get(this.PLANNER_API + '/cohorts')
      .map((res) => <Cohort[]>res.json());
  }

  findCohortByCode(code: string): Observable<Cohort> {
    return this._http.get(this.PLANNER_API + '/cohorts/' + code)
      .map((res: Response) => <Cohort>res.json());
  }

  saveCohort(cohort: Cohort): Observable<String> {
    return this._http.post(this.PLANNER_API + '/cohorts/' + cohort.code + '/save', JSON.stringify(cohort))
      .flatMap((res: Response) => Observable.of(res.text()))
      .catch((error) => this.handleError(error));
  }

  updateCohort(cohort: Cohort): Observable<String> {
    return this._http.put(this.PLANNER_API + '/cohorts/' + cohort.code, JSON.stringify(cohort))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeCohort(cohort: Cohort): Observable<Cohort> {
    return this._http.delete(this.PLANNER_API + '/cohorts/' + cohort.code)
      .map((res: Response) => <Cohort>res.json());
  }

  activateCohort(cohort: Cohort): Observable<String> {
    console.log('activate cohort');
    return this._http.get(this.PLANNER_API + '/cohorts/' + cohort.code + '/activate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  deactivateCohort(cohort: Cohort): Observable<String> {
    console.log('deactivate cohort');
    return this._http.get(this.PLANNER_API + '/cohorts/' + cohort.code + '/deactivate')
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // CURRICULUM
  // ====================================================================================================

  findCurriculums(): Observable<Curriculum[]> {
    console.log('findCurriculums');
    return this._http.get(this.PLANNER_API + '/curriculums')
      .map((res) => <Curriculum[]>res.json());
  }

  findCurriculumByCode(code: string): Observable<Curriculum> {
    return this._http.get(this.PLANNER_API + '/curriculums/' + code)
      .map((res: Response) => <Curriculum>res.json());
  }

  findSubjectsByCurriculum(curriculum: Curriculum): Observable<Subject[]> {
    console.log('findSubjectsByCurriculum');
    return this._http.get(this.PLANNER_API + '/curriculums/' + curriculum.code + '/subjects')
      .map((res: Response) => <Subject[]>res.json());
  }

  findSubjectsByCurriculumAndSubjectTypeCore(curriculum: Curriculum): Observable<Subject[]> {
    console.log('findSubjectsByCurriculumAndSubjectTypeCore');
    return this._http.get(this.PLANNER_API + '/curriculums/' + curriculum.code + '/subjects/subjectType/CORE')
      .map((res: Response) => <Subject[]>res.json());
  }

  findSubjectsByCurriculumAndSubjectTypeElective(curriculum: Curriculum): Observable<Subject[]> {
    console.log('findSubjectsByCurriculumAndSubjectTypeElective');
    return this._http.get(this.PLANNER_API + '/curriculums/' + curriculum.code + '/subjects/subjectType/ELECTIVE')
      .map((res: Response) => <Subject[]>res.json());
  }

  saveCurriculum(curriculum: Curriculum): Observable<String> {
    console.log(curriculum.code);
    return this._http.post(this.PLANNER_API + '/curriculums/' + curriculum.code + '/save', JSON.stringify(curriculum))
      .flatMap((res: Response) => Observable.of(res.text()))
      .catch((error) => this.handleError(error));
  }

  updateCurriculum(curriculum: Curriculum): Observable<String> {
    return this._http.put(this.PLANNER_API + '/curriculums/' + curriculum.code, JSON.stringify(curriculum))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  removeCurriculum(curriculum: Curriculum): Observable<Curriculum> {
    return this._http.delete(this.PLANNER_API + '/curriculums/' + curriculum.code)
      .map((res: Response) => <Curriculum>res.json());
  }

  // ====================================================================================================
  // SUBJECT
  // ====================================================================================================

  findSubjects(): Observable<Subject[]> {
    console.log('findSubjects');
    return this._http.get(this.PLANNER_API + '/subjects')
      .map((res) => <Subject[]>res.json());
  }

  addSingleSubject(curriculum: Curriculum, subject: SingleSubject): Observable<String> {
    console.log('curriculum:' + curriculum.code);
    return this._http.post(this.PLANNER_API + '/curriculums/' + curriculum.code + '/singleSubjects', JSON.stringify(subject))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  addBundleSubject(curriculum: Curriculum, subject: BundleSubject): Observable<String> {
    console.log('addSubject:' + subject);
    console.log('curriculum:' + curriculum.code);
    return this._http.post(this.PLANNER_API + '/curriculums/' + curriculum.code + '/bundleSubjects', JSON.stringify(subject))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

   addSubjectPart(curriculum: Curriculum, subject: BundleSubjectPart): Observable<String> {
    console.log('addSubjectPart:' + subject);
    console.log('curriculum:' + curriculum.code);
    return this._http.post(this.PLANNER_API + '/curriculums/' + curriculum.code + '/bundleSubjects', JSON.stringify(subject))
      .flatMap((res: Response) => Observable.of(res.text()));
  }


//   updateSubject(subject: Subject): Observable<String> {
//     return this._http.put(this.PLANNER_API + '/subjects/' + subject.id, JSON.stringify(subject))
//       .flatMap((res: Response) => Observable.of(res.text()));
//   }

  // findSubjectById(): Observable<Subject> {
  //   return this._http.get(this.PLANNER_API + '/subjects/' )
  //     .map((res: Response) => <Subject>res.json());
  // }

//  activateSubject(subject: Subject): Observable<String> {
//     console.log("activate subject");
//     return this._http.get(this.PLANNER_API + '/subjects/' + subject.id + '/activate')
//       .flatMap((res: Response) => Observable.of(res.text()));
//   }

  // deactivateSubject(subject: Subject): Observable<String> {
  //   console.log("deactivate cohort");
  //   return this._http.get(this.PLANNER_API + '/subjects/' + subject.id + '/deactivate')
  //     .flatMap((res: Response) => Observable.of(res.text()));
  // }

  // ====================================================================================================
  // PROGRAM LEVEL
  // ====================================================================================================

  findProgramLevels(): Observable<ProgramLevel[]> {
    return this._http.get(this.PLANNER_API + '/programLevels')
      .map((res: Response) => <ProgramLevel[]>res.json());
  }

  findProgramLevelByCode(code: string): Observable<ProgramLevel> {
    return this._http.get(this.PLANNER_API + '/programLevels/' + code)
      .map((res: Response) => <ProgramLevel>res.json());
  }

  activateProgramLevel(programLevel: ProgramLevel): Observable<String> {
    console.log('activate program level');
    return this._http.get(this.PLANNER_API + '/programLevels/' + programLevel.code + '/activate')
      .map((res: Response) => <String>res.json());
  }

  deactivateProgramLevel(programLevel: ProgramLevel): Observable<String> {
    console.log('deactivate program level');
    return this._http.get(this.PLANNER_API + '/programLevels/' + programLevel.code + '/deactivate')
      .map((res: Response) => <String>res.json());
  }

  saveProgramLevel(programLevel: ProgramLevel): Observable<String> {
    return this._http.post(this.PLANNER_API + '/programLevels/', JSON.stringify(programLevel))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  updateProgramLevel(programLevel: ProgramLevel): Observable<String> {
    return this._http.put(this.PLANNER_API + '/programLevels/' + programLevel.code, JSON.stringify(programLevel))
      .flatMap((res: Response) => Observable.of(res.text()));
  }

  // ====================================================================================================
  // PRIVATE METHODS
  // ====================================================================================================

  private handleError(error: Response | any) {
    let body: any = error.json();
    return Observable.throw(body);
  }

}

File: back/src/main/java/my/edu/umk/pams/academic/planner/service/PlannerService.java
package my.edu.umk.pams.academic.planner.service;

import my.edu.umk.pams.academic.identity.model.AdStudent;
import my.edu.umk.pams.academic.planner.model.*;
import my.edu.umk.pams.academic.web.module.planner.vo.AcademicYear;

import java.util.List;

/**
 * @author PAMS
 */
public interface PlannerService {

    //====================================================================================================
    // ACADEMIC SESSION
    //====================================================================================================

    AdAcademicSession findAcademicSessionById(Long id);

    AdAcademicSession findAcademicSessionByCode(String code);

    AdAcademicSession findCurrentAcademicSession();

    List<AdAcademicSession> findAcademicSessions(Integer offset, Integer limit);

    List<AdAcademicSession> findAcademicSessions(String filter, Integer offset, Integer limit);

	AdAcademicYear findByCode(String code);
    
    Integer countAcademicSession();

    Integer countAcademicSession(String filter);

    boolean isAcademicSessionCodeExists(String code);

    void saveAcademicSession(AdAcademicSession academicSession);

    void updateAcademicSession(AdAcademicSession academicSession);

    void removeAcademicSession(AdAcademicSession academicSession);

    //====================================================================================================
    // CURRICULUM
    //====================================================================================================

    AdCurriculum findCurriculumById(Long id);

    AdCurriculum findCurriculumByCode(String code);

    AdSubject findSubjectById(Long id);
    
    AdSubject findSubjectByCurriculum(AdCurriculum curriculum);

    AdSubject findSubjectByCurriculumAndCourse(AdCurriculum curriculum, AdCourse course);

    List<AdCurriculum> findCurriculums(Integer offset, Integer limit);

    List<AdCurriculum> findCurriculums(String filter, Integer offset, Integer limit);

    List<AdCurriculum> findCurriculums(AdProgram program);

    List<AdCurriculum> findCurriculums(AdProgram program, Integer offset, Integer limit);

    List<AdSubject> findSubjects(AdCurriculum curriculum);

    List<AdSubject> findSubjects(String filter, AdCurriculum curriculum);

    List<AdSubject> findSubjectByCourse(AdCourse course);

    Integer countCurriculum();

    Integer countCurriculum(String filter);

    Integer countSubject(AdCurriculum curriculum);

    boolean isCurriculumExists(String code);

    boolean isSubjectExists(AdCourse course, AdCurriculum curriculum);

    boolean hasCurriculum(AdProgram program);

    void saveCurriculum(AdCurriculum curriculum);

    void updateCurriculum(AdCurriculum curriculum);

    void removeCurriculum(AdCurriculum curriculum);

    void addSubject(AdCurriculum curriculum, AdSubject subject);

    void updateSubject(AdCurriculum curriculum, AdSubject subject);

    void deleteSubject(AdCurriculum curriculum, AdSubject subject);

    void addSubjectPart(AdBundleSubject subject, AdBundleSubjectPart part);

    void deleteSubjectPart(AdBundleSubject subject, AdBundleSubjectPart part);

    void addPrerequisites(AdCourse course, AdCourse... prerequisites);
    
    void updatePrerequisite(AdCourse course, AdCourse... prerequisites);
    
    void removePrerequisite(AdCourse course, AdCourse... prerequisites);

    //====================================================================================================
    // COHORT
    //====================================================================================================

    AdCohort findCohortByCode(String code);

    @Deprecated
    AdCohort findByParent(AdCohort parent);

    AdCohort findCohortById(Long id);

    List<AdCohort> findCohortRoots();

    List<AdCohort> findCohorts(String filter, Integer offset, Integer limit);

    List<AdCohort> findCohorts(AdCurriculum curriculum, Integer offset, Integer limit);

    List<AdCohort> findCohorts(AdProgram program, Integer offset, Integer limit);

    List<AdStudent> findStudents(AdCohort cohort);

    List<AdStudent> findStudents(AdCohort cohort, Integer offset, Integer limit);

    List<AdStudent> findStudents(String filter, AdCohort cohort, Integer offset, Integer limit);

    Integer countCohort(String filter);

    Integer countCohort(AdCohort parent);

    Integer countStudent(AdCohort cohort);

    Integer countCohort(AdProgram program);

    Integer countStudent(String filter, AdCohort cohort);

    boolean isCohortExists(String code);

    void saveCohort(AdCohort cohort);

    void updateCohort(AdCohort cohort);

    void removeCohort(AdCohort cohort);
    
    

    //====================================================================================================
    // FACULTY
    //====================================================================================================

    AdFaculty findFacultyById(Long id);

    AdFaculty findFacultyByCode(String code);

    List<AdFaculty> findFaculties();

    List<AdFaculty> findFaculties(Integer offset, Integer limit);

    List<AdFaculty> findFaculties(String filter, Integer offset, Integer limit);

    List<AdFaculty> findFaculties(boolean authorized, String filter, Integer offset, Integer limit);

    Integer countFaculty();

    Integer countFaculty(String filter);

    Integer countFaculty(boolean authorized, String filter);

    boolean isFacultyExists(String code);

    void saveFaculty(AdFaculty faculty);

    void updateFaculty(AdFaculty faculty);

    void removeFaculty(AdFaculty faculty);

    void addProgram(AdFaculty faculty, AdProgram program);

    void updateProgram(AdFaculty faculty, AdProgram program);

    void removeProgram(AdFaculty faculty, AdProgram program);

    void addCourse(AdFaculty faculty, AdCourse course);

    void updateCourse(AdFaculty faculty, AdCourse course);

    void removeCourse(AdFaculty faculty, AdCourse course);

    //====================================================================================================
    // INTAKE LEVEL
    //====================================================================================================

    AdProgramLevel findProgramLevelById(Long id);

    List<AdProgramLevel> findProgramLevels();

    List<AdProgramLevel> findProgramLevels(Integer offset, Integer limit);

    List<AdProgramLevel> findProgramLevels(String filter, Integer offset, Integer limit);

    Integer countProgramLevel();

    Integer countProgramLevel(String filter);

    void saveProgramLevel(AdProgramLevel level);

    void updateProgramLevel(AdProgramLevel level);

    void removeProgramLevel(AdProgramLevel level);

    
    //====================================================================================================
    // PROGRAM
    //====================================================================================================

    AdProgram findProgramById(Long id);

    AdProgram findProgramByCode(String code);

    AdProgram findProgramByCodeAndFaculty(String code, AdFaculty faculty);
    
    AdProgramLevel findProgramLevelByCode(String code);

    List<AdProgram> findPrograms(String filter);

    List<AdProgram> findPrograms(Integer offset, Integer limit);

    List<AdProgram> findPrograms(String filter, Integer offset, Integer limit);

    List<AdProgram> findPrograms(AdFaculty faculty);

    List<AdProgram> findPrograms(AdProgramType type, AdFaculty faculty);

    List<AdProgram> findPrograms(AdFaculty faculty, Integer offset, Integer limit);

    List<AdProgram> findPrograms(AdProgramType type, AdFaculty faculty, Integer offset, Integer limit);

    List<AdProgram> findPrograms(String filter, AdFaculty faculty, Integer offset, Integer limit);

    List<AdProgram> findPrograms(String filter, AdProgramType type, AdFaculty faculty, Integer offset, Integer limit);

    Integer countProgram();

    Integer countProgram(String filter);

    Integer countProgram(String filter, AdFaculty faculty);

    Integer countProgram(String filter, AdProgramType type, AdFaculty faculty);

    Integer countProgram(AdFaculty faculty);

    Integer countProgram(AdProgramType type, AdFaculty faculty);

    boolean isProgramExists(String code, AdFaculty faculty);
    
    void saveProgram(AdProgram program);

    void updateProgram(AdProgram program);

    void removeProgram(AdProgram program);

    //====================================================================================================
    // COURSE
    //====================================================================================================
    AdCourse findCourseById(Long id);

    AdCourse findCourseByCode(String code);

    AdCourse findCourseByCodeAndFaculty(String code, AdFaculty faculty);

    List<AdCourse> findAvailableCourses(AdAcademicSession academicSession, Integer offset, Integer limit);

    List<AdCourse> findCourses(Integer offset, Integer limit);

    List<AdCourse> findCourses(String filter, Integer offset, Integer limit);

    List<AdCourse> findCourses(AdFaculty faculty);

    List<AdCourse> findCourses(String filter, AdFaculty faculty, Integer offset, Integer limit);

    Integer countCourse();

    Integer countCourse(String filter);

    Integer countCourse(AdFaculty faculty);

    Integer countCourse(String filter, AdFaculty faculty);

    Integer countCourse(boolean authorized, String filter);

    Integer countAvailableCourse(AdAcademicSession academicSession);

    boolean isCourseExists(String code, AdFaculty faculty);
    
    void saveCourse(AdCourse course , AdFaculty faculty);

    void updateCourse(AdCourse course);

    void removeCourse(AdCourse course);

 
  //====================================================================================================
    // ACADEMIC YEAR
    //====================================================================================================
    List<AdAcademicYear> findAcademicYears(Integer offset, Integer limit);

    List<AdAcademicYear> findAcademicYears(String filter, Integer offset, Integer limit);
    
    void saveAcademicYear(AdAcademicYear year);

}
